data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/S3SinkService.java

```diff
@@ -0,0 +1,130 @@
+import org.opensearch.dataprepper.plugins.sink.accumulator.Buffer;
+import org.opensearch.dataprepper.plugins.sink.accumulator.BufferFactory;
+import org.opensearch.dataprepper.plugins.sink.accumulator.ObjectKey;
+import org.opensearch.dataprepper.plugins.sink.codec.Codec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration;
+import software.amazon.awssdk.core.retry.RetryPolicy;
+import software.amazon.awssdk.services.s3.S3Client;
+
+/**
+ * Class responsible for create {@link S3Client} object, check thresholds,
+ * get new buffer and write records into buffer.
+ */
+public class S3SinkService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(S3SinkService.class);
+    private final S3SinkConfig s3SinkConfig;
+    private final Lock reentrantLock;
+    private final BufferFactory bufferFactory;
+    private final Codec codec;
+    private final S3Client s3Client;
+    private Buffer currentBuffer;
+    private final int maxEvents;
+    private final ByteCount maxBytes;
+    private final long maxCollectionDuration;
+
+    /**
+     * @param s3SinkConfig s3 sink related configuration.
+     * @param codec        parser
+     */
+    public S3SinkService(final S3SinkConfig s3SinkConfig, final BufferFactory bufferFactory, final Codec codec) {
+        this.s3SinkConfig = s3SinkConfig;
+        this.bufferFactory = bufferFactory;
+        this.codec = codec;
+        this.s3Client = createS3Client();
+        reentrantLock = new ReentrantLock();
+
+        maxEvents = s3SinkConfig.getThresholdOptions().getEventCount();
+        maxBytes = s3SinkConfig.getThresholdOptions().getMaximumSize();
+        maxCollectionDuration = s3SinkConfig.getThresholdOptions().getEventCollectTimeOut().getSeconds();
+    }
+
+    /**
+     * @param records received records and add into buffer.
+     */
+    void output(Collection<Record<Event>> records) {
+        reentrantLock.lock();
+        final String bucket = s3SinkConfig.getBucketOptions().getBucketName();
+        if (currentBuffer == null) {
+            currentBuffer = bufferFactory.getBuffer();
+        }
+        try {
+            for (Record<Event> record : records) {
+
+                final Event event = record.getData();
+                final String encodedEvent;
+                encodedEvent = codec.parse(event);
+                final byte[] encodedBytes = encodedEvent.getBytes();
+                if (willExceedThreshold()) {
+                    LOG.info("Snapshot info : Byte_capacity = {} Bytes," +
+                            " Event_count = {} Records & Event_collection_duration = {} Sec",
+                            maxBytes.getBytes(), currentBuffer.getEventCount(), currentBuffer.getDuration());
+                    boolean isUploadedToS3 = currentBuffer.flushToS3(s3Client, bucket, generateKey());
+                    if (isUploadedToS3) {
```

Comment:
We should retry few times before failing. And log the failure reason clearly.

Reply:
Resolved.